### ISSUE TITLE
using  files in template parent directories to create files in generated project directories

### ISSUE THREAD

**Issue by wt** [cargo-generate/cargo-generate#1054]
I have a git repo with multiple generation templates.

```
<root>/examples/
-> aaa
-> bbb
```

Some of the templates have files with the same contents.

```
<root>examples/aaa
-> Makefile.toml (uses ../cargo-make/main.toml)
```

```
<root>examples/bbb
-> Makefile.toml (uses ../cargo-make/main.toml)
```

I can't find a way to use the common files to create files in the generated template.

I would like to be able to do something like.

```
[template]
include_common = [
  {path="../cargo-make", dest_path="cargo-make"},
]
```

I want this to have the effect of using the files in the `../cargo-make/` as templates for corresponding files that are created in the `cargo-make` subfolder of the generated project root.

Even rhai scripts only appear to be able to rename files in the generated folder or generate files from a string or array into the generated folder. The rhai scripts don't appear to allow reading a file within a parent directory of the template directory.

Is there a way to do what I want with cargo-generate right now that I am not seeing?

Thanks a bunch!

**Comment by taurr** [comment]
Nope, `cargo-generate` restricts access to files residing within the template directory for security reasons.

It could be argued that the rule should be relaxed to allow access within the repository, but that would be difficult to ensure since templates can also come from raw folders. Therefore, the template folder represents the broadest scope that is allowed.

**Comment by wt** [comment]
How would you recommend handling multiple templates in a single repo that all refer to files that contain the same contents.

Here's an example of a bunch of examples that could also be templates if not for the shared files: https://github.com/leptos-rs/leptos/tree/main/examples

**Comment by MabezDev** [comment]
I am also looking for something like this, in esp-rs, we will have many templates that share many of the same files, only some parts in main.rs and Cargo.toml changing. I really don't want to be duplicating a bunch of these files, across the repo.

> It could be argued that the rule should be relaxed to allow access within the repository, but that would be difficult to ensure since templates can also come from raw folders

Could you expand on this a little? I don't quite understand the connection from repo root to raw folders, are they not exclusive? I.e when calling cargo generate its either a repo via `--git`, or a raw folder via `--path`. Could we at least relax the rules in the `--git` case?

If that isn't possible/suitable, do you have any other suggestions to achieve this? I'm happy to work on a PR.

**Comment by atty303** [comment]
I also want to place shared files in subprojects of a Git repository. For example, I think this could be solved with the following approach.

1. Expand the sandbox up to the Git repository root. If `git rev-parse --show-toplevel` succeeds, use that as the sandbox; if it fails because it’s not a Git repository, keep the current behavior and use the template directory as the sandbox.
2. Allow symlinks and commit them to Git as symlinks. Currently, symlinks are explicitly skipped and don’t work. However, since it’s sandboxed, I think it’s safe to resolve the symlink.

Below is a fairly heavy-handed workaround. I’m using it for my personal templates, but I wouldn’t recommend it for templates intended for distribution. For distributable templates, I think the best approach for now is to use a GitHub Action in the template repository (or similar) to copy the shared files into the subprojects.

### Workaround using `system::command`

First, write the files you want to copy into the init hook of the subproject.

`/sub/cargo-generate.toml`:
```toml
[template]
ignore = ["cargo-generate-mod.rhai"]

[hooks]
init = ["cargo-generate-init.rhai"]
```

`/sub/cargo-generate-init.rhai`:
```
import "cargo-generate-mod" as mod;

mod::copy_files("copy-shared-files", "../shared", [
  "hello.txt",
]);
```

Generate a shell script from the file list and execute it.

`/cargo-generate-mod.rhai`:
```
fn copy_files(name, base, files) {
  let is_windows = variable::get("os-arch").starts_with("windows");

  let script = [];
  for file in files {
    let source = `${base}/${file}`;
    if is_windows {
      script.push(`New-Item -ItemType Directory -Force -Path (Split-Path -Parent '${file}') | Out-Null`);
      script.push(`Copy-Item '${source}' '${file}'`);
    } else {
      script.push(`mkdir -p $(dirname "${file}") && cp -p "${source}" "${file}"`);
    }
  }

  if is_windows {
    let script_file = `${name}.ps1`;
    file::write(script_file, script);
    system::command("powershell", ["-File", script_file]);
    file::delete(script_file);
  } else {
    let script_file = `${name}.sh`;
    file::write(script_file, script);
    system::command("sh", ["-e", script_file]);
    file::delete(script_file);
  }
}
```

Because of the sandbox, `cargo-generate-mod.rhai` cannot be loaded unless the actual file is placed in the subproject.

```bash
cp cargo-generate-mod.rhai sub/cargo-generate-mod.rhai

# You can update all scripts
find . -mindepth 2 -name cargo-generate-mod.rhai -exec cp cargo-generate-mod.rhai {} \;
```
